### PR TITLE
Add an extension method Users to IHubClients witch takes IEnumerable parameter

### PR DESCRIPTION
**Pull request by Marusyk**
Add a simple extension method toIHubClients witch takes IEnumerable parameter

Addresses #17285

Please review
Thank you in advance

### CODE REVIEW

**Comment by anurse** [comment]
We should add this to all the client methods if we're going to do it (Groups, GroupsExcept, etc. etc. etc.). Also, we need to discuss if an extension method is right or if we should use a Default Interface Member. The latter would allow an implementation to override it and provide different logic for `IEnumerable` over `IReadOnlyList`.

Thoughts on this @davidfowl @halter73 @BrennanConroy ?

**Comment by halter73** [comment]
I agree we should do the same for Groups, GroupsExcept, etc... I think for something as trivial as this, and extension method is still appropriate. I don't know why an interface implementor would want to do anything special for `IEnumerable` vs `IReadOnlyList`

**Comment by Marusyk** [comment]
Can anybody suggest how to fix build and what is Pull Request Labeler?

**Comment by halter73** [comment]
> Can anybody suggest how to fix build and what is Pull Request Labeler?

https://github.com/aspnet/AspNetCore/blob/master/docs/ReferenceAssemblies.md#when-changing-public-api

@aspnet/asp-net-api-reviews We'll want to do an API review on this before merging.

I wouldn't worry about the PR labeler. That doesn't block the build. I don't even think it's supposed to run for external PRs (https://github.com/aspnet/AspNetCore/pull/15171) @mkArtakMSFT Do you know what's going on here?

**Comment by Marusyk** [comment]
Thanks,

I cann't build localy because https://github.com/aspnet/AspNetCore/blob/master/docs/BuildFromSource.md#common-error-unable-to-locate-the-net-core-sdk

**Comment by anurse** [comment]
> I can't build locally because [/docs/BuildFromSource.md@`master`#common-error-unable-to-locate-the-net-core-sdk](https://github.com/aspnet/AspNetCore/blob/master/docs/BuildFromSource.md#common-error-unable-to-locate-the-net-core-sdk) [](/aspnet/AspNetCore/blob/HEAD@{2019-12-04T22:19:23Z}/docs/BuildFromSource.md#common-error-unable-to-locate-the-net-core-sdk)

Have you enabled the "Use previews of the .NET Core SDK" checkbox as suggested in that doc?

Even if you can't fully build, you may still be able to run the script @halter73 mentioned to update the ref assembly. Updating that is necessary to resolve the Code Check failure.

**Comment by Marusyk** [comment]
@anurse yes, I have enabled the "Use previews of the .NET Core SDK" checkbox as suggested in that doc

**Comment by anurse** [comment]
Excellent! Thanks!

A note here about the process from here:

1. Someone from the team needs to review this and sign off (ping @halter73 @BrennanConroy  ;)).

1. We'll review the new API in our regular API review meeting (we do this for all public API changes in mature areas like SignalR) and either approve it or provide further feedback.

Once we've got those done, we'll merge it in!

File: src/SignalR/server/Core/src/HubClientsExtensions.cs [hunk 1]
```diff
@@ -482,6 +516,17 @@ public static T GroupExcept<T>(this IHubClients<T> hubClients, string groupName,
             return hubClients.GroupExcept(groupName, new [] { excludedConnectionId1, excludedConnectionId2, excludedConnectionId3, excludedConnectionId4, excludedConnectionId5, excludedConnectionId6, excludedConnectionId7, excludedConnectionId8 });
         }
 
+        /// <summary>
+        /// Gets a <typeparamref name="T" /> that can be used to invoke methods on all connections in the specified group excluding the specified connections.
+        /// </summary>
+        /// <param name="hubClients">The abstraction that provides access to connections.</param>
+        /// <param name="excludedConnectionIds">The connection IDs to exclude.</param>
+        /// <returns>A <typeparamref name="T" /> representing the methods that can be invoked on the clients.</returns>
+        public static T GroupExcept<T>(this IHubClients<T> hubClients, IEnumerable<string> excludedConnectionIds)
```
Comment (BrennanConroy) [inline src/SignalR/server/Core/src/HubClientsExtensions.cs]: This is missing the group name

**Review by BrennanConroy — CHANGES_REQUESTED** [review]

**Review by BrennanConroy — APPROVED** [review]

**Review by halter73 — APPROVED** [review]

**Review by pranavkm — COMMENTED** [review]
We want to make sure this isn't a source breaking change

**Comment by rynowak** [comment]
> 
> 
> We want to make sure this isn't a source breaking change

Specifically we should make sure that using array or list here doesn't lead to an ambiguity compiler diagnostic.

**Comment by halter73** [comment]
I tried this with an Array, List and IEnumerable (which was really an Array stored as an IEnumerable). Since both Array and List implement IReadOnlyList and IReadOnlyList inherits from IEnumerable, the compiler chooses the more-derived IReadOnlyList overload. In the IEnumerable case, the compiler chooses the IEnumerable extension method since that's the only workable overload.

Overload selection is a compile-time decision, so an Array/List/whatever can still be passed to the IEnumerable extension method if the compiler only knows it's an IEnumerable, but I don't see a problem with that.

**Comment by halter73** [comment]
Thanks @Marusyk.

**Pull request merged**